tests/unit/s2n_x509_validator_custom_oids_test.c

```diff
@@ -0,0 +1,208 @@
+                .key_pem_path = S2N_MULTIPLE_OID_KEY,
+                .custom_critical_oids = single_oid,
+                .custom_oid_count = 1,
+                .set_oids = true,
+                .expected_error = S2N_ERR_CERT_UNTRUSTED,
+            },
+        };
+        /* clang-format on */
+
+        /* Self-talk: add custom critical extensions for server auth */
+        for (int i = 0; i < s2n_array_len(test_cases); i++) {
+            DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+            EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                    test_cases[i].cert_pem_path, test_cases[i].key_pem_path));
+
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+            EXPECT_SUCCESS(s2n_config_set_verification_ca_location(config, test_cases[i].cert_pem_path, NULL));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+
+            const char **custom_oids = test_cases[i].custom_critical_oids;
+            uint32_t custom_oid_count = test_cases[i].custom_oid_count;
+
+            EXPECT_NULL(config->custom_crit_oids);
+            if (test_cases[i].set_oids) {
+                EXPECT_SUCCESS(s2n_config_set_custom_critical_oids(config, custom_oids, custom_oid_count));
+                EXPECT_EQUAL(sk_ASN1_OBJECT_num(config->custom_crit_oids), custom_oid_count);
+            }
+
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_set_server_name(client_conn, "localhost"));
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+            s2n_error expected_error = test_cases[i].expected_error;
+            if (expected_error == S2N_ERR_OK) {
+                EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            } else {
+                EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                        expected_error);
```

Comment:
`S2N_ERR_CERT_UNTRUSTED` is a very generic error. To make sure the libcrypto is failing for the reason you think it is, it might be good to check the actual error code from the libcrypto ([X509_STORE_CTX_get_error](https://docs.openssl.org/3.0/man3/X509_STORE_CTX_get_error/))